src/librbd/Operations.cc

```diff
@@ -1486,19 +1486,18 @@ int Operations<I>::metadata_set(const std::string &key,
     return -EROFS;
   }
 
-  C_SaferCond metadata_ctx;
-  {
-    std::shared_lock owner_lock{m_image_ctx.owner_lock};
-    r = prepare_image_update(exclusive_lock::OPERATION_REQUEST_TYPE_GENERAL,
-                             true);
-    if (r < 0) {
-      return r;
-    }
-
-    execute_metadata_set(key, value, &metadata_ctx);
+  r = invoke_async_request("metadata_set",
+                           exclusive_lock::OPERATION_REQUEST_TYPE_GENERAL,
+                           true,
+                           boost::bind(&Operations<I>::execute_metadata_set,
+                                       this, key, value, _1),
+                           boost::bind(&ImageWatcher<I>::notify_metadata_set,
+                                       m_image_ctx.image_watcher,
+                                       key, value, _1));
+  if (r < 0 && r != -EINVAL) {
```

Comment:
Nit: does the `metadata_set` operation ever return `EINVAL`?

Reply:
Yes, it will return `EINVAL` if failed to decoding `buf`.
https://github.com/ceph/ceph/blob/6b31e3747be7c9831dd49f0fe81346bca7baac99/src/cls/rbd/cls_rbd.cc#L3792

Reply:
Well, that's not really an expected error caused by replays. We only want to filter out errors that could occur if the same RPC method was invoked twice (i.e. a double remove attempt shouldn't return -ENOENT on the second call if the first call succeeded).

Reply:
@dillaman Got it and repushed, please take a look. Thank you.